### ISSUE TITLE
Allow the block icon and block title to be dynamic

### ISSUE THREAD

**Issue by bfintal** [WordPress/gutenberg#51776]
## What problem does this address?
Allow the **block icon** and **block title** to be changed dynamically after registration. This can be for visual and UX purposes only and allowing these values to be dynamic would change the icon and title in various places such as the inspector card, list view icon & title, and toolbar icon.

Just some background and context: In my block plugin that provides third-party blocks, I have an "Inner Column" block that is reused as an inner block in a number of different parent blocks. For example:

- Columns block > Inner Column block
- Carousel block > Inner Column block
- Tabs block > Tab Content block > Inner Column block

Here's what my List View looks like:
![image](https://github.com/WordPress/gutenberg/assets/1033611/009cb6d3-aa99-4c00-b5b8-f924fe9a248c)

While this works, it would be better for the user's experience if the "Inner Column" blocks were changed into something more appropriate to the parent. Technically, I could create new blocks to address this, but I'm not sure registering a completely new block with just the title and icon change would be an elegant solution.

Now to improve the user experience, it would be better if they saw **Tab Panel blocks** and **Slide blocks** instead (with corresponding icon changes):
![image](https://github.com/WordPress/gutenberg/assets/1033611/6c41f20d-f1ce-4c22-8b1b-c8327f9ed546)

## What is your proposed solution?
I propose we alternatively allow a callback to be used in the `title` and `icon` parameters when registering a block. These callbacks can be used to dynamically generate the icon and title when needed.

```js
// Current way.
registerBlockType( 'my-plugin/inner-column', {
    title: __( 'Inner Column', 'textdomain' ),
} );

// What if we also allow a callback.
registerBlockType( 'my-plugin/inner-column', {
    title: () => __( 'Inner Column', 'textdomain' ),
} );
```

Arguments can be passed to the callback function to help out in determining the title and icon. In the scenario I gave, hooks could be used to grab the context / parent block name and output a different icon and title accordingly.

**Comment by talldan** [comment]
@bfintal Using block variations might be one option. Though I'm not sure if variations can be limited as the children of a specific block.

**Comment by bfintal** [comment]
> @bfintal Using block variations might be one option. Though I'm not sure if variations can be limited as the children of a specific block.

Good idea! I'm a bit weary of adding a bunch of hidden variations each with their own className just for this, but I tried it and it worked:

```js
variations: [
	{
		name: 'slide',
		title: __( 'Slide', 'textdomain' ),
		attributes: {
			className: 'is-style-slide',
		},
		isActive: [ 'className' ],
		scope: [],
	},
],
```
